File: packages/gemma/src/verifier/scope-enforcer.ts
import { ScopeViolationError } from '../errors.js';

/**
 * Checks whether `grantScopes` contains a specific scope string.
 */
export function hasScope(grantScopes: string[], scope: string): boolean {
  return grantScopes.includes(scope);
}

/**
 * Throws {@link ScopeViolationError} if any of `requiredScopes` is missing
 * from `grantScopes`.
 */
export function enforceScopes(
  grantScopes: string[],
  requiredScopes: string[],
): void {
  const missing = requiredScopes.filter((s) => !grantScopes.includes(s));
  if (missing.length > 0) {
    throw new ScopeViolationError(requiredScopes, grantScopes);
  }
}

File: packages/gemma-py/src/grantex_gemma/_scope_enforcer.py
"""Scope checking utilities."""

from __future__ import annotations

from ._errors import ScopeViolationError


def enforce_scopes(
    grant_scopes: list[str],
    required_scopes: list[str],
) -> None:
    """Ensure all required scopes are present in the grant.

    Raises ScopeViolationError if any required scope is missing.
    """
    missing = [s for s in required_scopes if s not in grant_scopes]
    if missing:
        raise ScopeViolationError(
            f"Missing required scopes: {', '.join(missing)}"
        )


def has_scope(grant_scopes: list[str], scope: str) -> bool:
    """Check whether a single scope is present in the grant."""
    return scope in grant_scopes
